File: Estructura de datos/Arrays Abstractos/ArrayAbstracto.java
abstract class PersonaBase {
    public abstract double imc();
    public abstract boolean mayorEdad();
    public abstract void info();
}

class Persona extends PersonaBase {
    private String nom;
    private int edad;
    private double kg;
    private double mts;
    private String mail;
    private String cel;

    public Persona(String n, int e, double p, double a) {
        this(n, e, p, a, null, null);
    }

    public Persona(String n, int e, double p, double a, String m, String c) {
        nom = n;
        edad = e;
        kg = p;
        mts = a;
        mail = m;
        cel = c;
    }

    @Override
    public double imc() {
        return kg / (mts * mts);
    }

    @Override
    public boolean mayorEdad() {
        return edad >= 18;
    }

    @Override
    public void info() {
        System.out.print(nom + ", " + edad + " años, IMC: " + String.format("%.2f", imc()));
        if (mail != null) System.out.print(", Email: " + mail);
        if (cel != null) System.out.print(", Cel: " + cel);
        System.out.println();
    }

    public String estadoIMC() {
        double v = imc();
        if (v < 18.5) return "Bajo peso";
        else if (v < 25) return "Normal";
        else if (v < 30) return "Sobrepeso";
        else return "Obesidad";
    }
}

public class Main {
    public static void main(String[] args) {
        System.out.println("=== Demo Persona Java ===");

        Persona p1 = new Persona("Ivan", 25, 70.5, 1.75, "elparadero@gmail.com", null);
        p1.info();
        System.out.println("Mayor de edad? " + (p1.mayorEdad() ? "Sí" : "No"));
        System.out.println("IMC: " + p1.estadoIMC());

        Persona p2 = new Persona("Helmer", 17, 65, 1.68, "ivan_lizarraga@gmail.com", "123456");
        p2.info();
        System.out.println("Mayor de edad? " + (p2.mayorEdad() ? "Sí" : "No"));

        PersonaBase p3 = new Persona("Jesus", 30, 80, 1.80);
        p3.info();
        System.out.println("IMC calc: " + String.format("%.2f", p3.imc()));
    }
}

File: Estructura de datos/Arrays Abstractos/ArrayAbstracto.cpp
#include <iostream>
#include <string>
#include <optional>
#include <memory>
#include <iomanip>

class PersonaBase {
public:
    virtual ~PersonaBase() = default;
    virtual double imc() const = 0;
    virtual bool mayorEdad() const = 0;
    virtual void info() const = 0;
};

class Persona : public PersonaBase {
private:
    std::string nom;
    int edad;
    double kg;
    double mts;
    std::optional<std::string> correo;
    std::optional<std::string> cel;

public:
    Persona(std::string n, int e, double p, double a,
            std::optional<std::string> mail = std::nullopt,
            std::optional<std::string> t = std::nullopt)
        : nom(std::move(n)), edad(e), kg(p), mts(a), correo(std::move(mail)), cel(std::move(t)) {}

    double imc() const override { return kg / (mts * mts); }
    bool mayorEdad() const override { return edad >= 18; }

    void info() const override {
        std::cout << nom << ", " << edad << " años, IMC: "
                  << std::fixed << std::setprecision(2) << imc();
        if (correo) std::cout << ", Email: " << *correo;
        if (cel) std::cout << ", Cel: " << *cel;
        std::cout << "\n";
    }

    std::string estadoIMC() const {
        double v = imc();
        if (v < 18.5) return "Bajo peso";
        else if (v < 25) return "Normal";
        else if (v < 30) return "Sobrepeso";
        else return "Obesidad";
    }

    void setEdad(int e) { edad = e; }
    void setPeso(double p) { kg = p; }
    void setAltura(double a) { mts = a; }
};

void probar() {
    std::cout << "=== Demo Persona ===\n";

    Persona p1("Ivan", 25, 70.5, 1.75, "elparadero@gmail.com");
    p1.info();
    std::cout << "Mayor? " << (p1.mayorEdad() ? "Sí" : "No") << "\n";
    std::cout << "IMC: " << p1.estadoIMC() << "\n";

    Persona p2("Helmer", 17, 65, 1.68, "ivan_lizarraga@gmail.com", "123456");
    p2.info();
    std::cout << "Mayor? " << (p2.mayorEdad() ? "Sí" : "No") << "\n";
}

void probarSmart() {
    std::unique_ptr<PersonaBase> per = std::make_unique<Persona>("Jesus", 30, 80, 1.80);
    per->info();
    std::cout << "IMC calc: " << per->imc() << "\n";
}

int main() {
    probar();
    std::cout << "\n=== Con SmartPointers ===\n";
    probarSmart();
    return 0;
}
